src/theory/bags/bags_utils.cpp

```diff
@@ -785,6 +787,95 @@ Node BagsUtils::evaluateBagFold(TNode n)
   return ret;
 }
 
+Node BagsUtils::evaluateBagPartition(Rewriter* rewriter, TNode n)
+{
+  Assert(n.getKind() == BAG_PARTITION);
+  NodeManager* nm = NodeManager::currentNM();
+
+  // Examples
+  // --------
+  // minimum string
+  // - (bag.partition
+  //     ((lambda ((x Int) (y Int)) (= 0 (+ x y)))
+  //     (bag.union_disjoint
+  //       (bag 1 20) (bag (- 1) 50)
+  //       (bag 2 30) (bag (- 2) 60)
+  //       (bag 3 40) (bag (- 3) 70)
+  //       (bag 4 100)))
+  //   = (bag.union_disjoint
+  //       (bag (bag 4 100) 1)
+  //       (bag (bag.union_disjoint (bag 1 20) (bag (- 1) 50)) 1)
+  //       (bag (bag.union_disjoint (bag 2 30) (bag (- 2) 60)) 1)
+  //       (bag (bag.union_disjoint (bag 3 40) (bag (- 3) 70)) 1)))
+
+  Node r = n[0];  // equivalence relation
+  Node A = n[1];  // bag
+  TypeNode bagType = A.getType();
+  TypeNode partitionType = n.getType();
+  std::map<Node, Rational> elements = BagsUtils::getBagElements(A);
+  Trace("bags-partition") << "elements: " << elements << std::endl;
+  // a simple map from elements to equivalent classes with this invariant:
+  // each key element must appear exactly once in one of the values.
+  std::map<Node, std::set<Node>> sets;
+  std::set<Node> emptyClass;
+  for (const auto& pair : elements)
+  {
+    // initially each singleton element is an equivalence class
+    sets[pair.first] = {pair.first};
+  }
+  for (std::map<Node, Rational>::iterator i = elements.begin();
+       i != elements.end();
+       ++i)
+  {
+    if (sets[i->first].empty())
+    {
+      // skip this element since its equivalent class has already been processed
+      continue;
+    }
+    std::map<Node, Rational>::iterator j = i;
+    ++j;
+    while (j != elements.end())
+    {
+      Node sameClass = nm->mkNode(APPLY_UF, r, i->first, j->first);
+      sameClass = rewriter->rewrite(sameClass);
+      Assert(sameClass.isConst())
```

Comment:
I think we definitely shouldn't expect that this will always be constant.

I think perhaps we should make this a conditional and then exit, returning `n` itself here.

Reply:
Done. The map that needs to be partitioned must be constant for this function. The problem is with the equivalent relation, which could be symbolic and lambda definition.